File: backend/src/json_rpc/protocol.py
from enum import Enum, auto
import json


class MESSAGE_TYPE(Enum):
    REQUEST = auto()
    NOTIFICATION = auto()
    RESPONSE = auto()
    ERROR = auto()


class ERROR_TYPE(Enum):
    PARSE_ERROR = -32700
    INVALID_REQUEST = -32600
    METHOD_NOT_FOUND = -32601
    INVALID_PARAMS = -32602
    INTERNAL_ERROR = -32603


class JsonRpcError(Exception):
    ERROR_TYPE = ERROR_TYPE.INTERNAL_ERROR
    ERROR_MESSAGE = 'Internal error'

    def get_error_message(self):
        if self.args:
            return self.args[0]

        return self.ERROR_MESSAGE


class InvalidRequestError(JsonRpcError):
    ERROR_TYPE = ERROR_TYPE.INVALID_REQUEST
    ERROR_MESSAGE = 'Invalid request'


class InvalidParamsError(JsonRpcError):
    ERROR_TYPE = ERROR_TYPE.INVALID_PARAMS
    ERROR_MESSAGE = 'Invalid params'


def decode_message(message_string):
    '''
    returns (error_type, error_message, message_type, message)
    '''

    try:
        message = json.loads(message_string)

    except json.JSONDecodeError:

        # raw message is no valid JSON
        return ERROR_TYPE.PARSE_ERROR, 'Parse error', None, None

    # every JSON RPC message has to contain {"jsonrpc": "2.0"}
    if 'jsonrpc' not in message or message['jsonrpc'] != '2.0':
        return (
            ERROR_TYPE.INVALID_REQUEST,
            "Missing required 'jsonrpc' field.",
            None,
            None,
        )

    message_type = None

    # request / notification
    if 'method' in message:

        # methods have to be strings
        if not isinstance(message['method'], str):
            return (
                ERROR_TYPE.INVALID_REQUEST,
                'Method has to be a string.',
                None,
                None,
            )

        # request
        if 'id' in message:

            # ids have to be a string, number
            if not isinstance(message['id'], (str, int, float)):
                return (
                    ERROR_TYPE.PARSE_ERROR,
                    'Invalid id field: must be a string, number, or null.',
                    None,
                    None,
                )

            message_type = MESSAGE_TYPE.REQUEST

        # notification
        else:
            message_type = MESSAGE_TYPE.NOTIFICATION

    # response
    elif 'result' in message:

        # every result must have an id
        if 'id' not in message:
            return (
                ERROR_TYPE.PARSE_ERROR,
                'Missing id field.',
                None,
                None,
            )

        # ids have to be a string, number
        if not isinstance(message['id'], (str, int, float)):
            return (
                ERROR_TYPE.PARSE_ERROR,
                'Invalid id field: must be a string, number, or null.',
                None,
                None,
            )

        message_type = MESSAGE_TYPE.RESPONSE

    # error
    elif 'error' in message:
        message_type = MESSAGE_TYPE.ERROR

    # valid message
    return None, "", message_type, message


def encode_notification(method, params=None):
    return json.dumps({
        'jsonrpc': '2.0',
        'method': method,
        'params': params,
    })


def encode_result(message_id, result):
    return json.dumps({
        'jsonrpc': '2.0',
        'id': message_id,
        'result': result,
    })


def encode_error(error_type, error_message, message_id=None, data=None):
    message = {
        'jsonrpc': '2.0',
        'error': {
            'code': error_type.value,
            'message': error_message,
        },
    }

    if message_id is not None:
        message['id'] = message_id

    if data:
        message['error']['data'] = data

    return json.dumps(message)

File: frontend/src/json-rpc/protocol.ts
export enum MESSAGE_TYPE {
  REQUEST = "REQUEST",
  NOTIFICATION = "NOTIFICATION",
  RESPONSE = "RESPONSE",
  ERROR = "ERROR",
}

export enum ERROR_TYPE {
  PARSE_ERROR = -32700,
  INVALID_REQUEST = -32600,
  METHOD_NOT_FOUND = -32601,
  INVALID_PARAMS = -32602,
  INTERNAL_ERROR = -32603,
}

export type JsonRpcMessagePayload = {
  jsonrpc: string;
  id?: number | string | null;
  method?: string;
  params?: unknown;
  result?: unknown;
  error?: {
    code: number;
    message: string;
    data?: unknown;
  };
};

export type JsonRpcMessage = {
  errorType: ERROR_TYPE | null;
  errorMessage: string;
  messageType: MESSAGE_TYPE | null;
  message: JsonRpcMessagePayload;
};

export function decodeMessage(messageString: string): JsonRpcMessage {
  let message: JsonRpcMessagePayload;

  try {
    message = JSON.parse(messageString) as JsonRpcMessagePayload;
  } catch {
    // raw message is no valid JSON
    return {
      errorType: ERROR_TYPE.PARSE_ERROR,
      errorMessage: "Parse error",
      messageType: null,
      message: null,
    };
  }

  // every JSON RPC message has to contain {"jsonrpc": "2.0"}
  if (message.jsonrpc !== "2.0") {
    return {
      errorType: ERROR_TYPE.INVALID_REQUEST,
      errorMessage: "Missing required 'jsonrpc' field.",
      messageType: null,
      message: null,
    };
  }

  let messageType: MESSAGE_TYPE | null = null;

  // request / notification
  if ("method" in message) {
    // methods have to be strings
    if (typeof message.method !== "string") {
      return {
        errorType: ERROR_TYPE.INVALID_REQUEST,
        errorMessage: "Method has to be a string.",
        messageType: null,
        message: null,
      };
    }

    // request
    if ("id" in message) {
      // ids have to be a string, number
      if (
        typeof message.id !== "string" &&
        typeof message.id !== "number" &&
        message.id !== null
      ) {
        return {
          errorType: ERROR_TYPE.PARSE_ERROR,
          errorMessage: "Invalid id field: must be a string, number, or null.",
          messageType: null,
          message: null,
        };
      }

      messageType = MESSAGE_TYPE.REQUEST;

      // notification
    } else {
      messageType = MESSAGE_TYPE.NOTIFICATION;
    }

    // response
  } else if ("result" in message) {
    // every result must have an id
    if (!("id" in message)) {
      return {
        errorType: ERROR_TYPE.PARSE_ERROR,
        errorMessage: "Missing id field.",
        messageType: null,
        message: null,
      };
    }

    // ids have to be a string, number
    if (
      typeof message.id !== "string" &&
      typeof message.id !== "number" &&
      message.id !== null
    ) {
      return {
        errorType: ERROR_TYPE.PARSE_ERROR,
        errorMessage: "Invalid id field: must be a string, number, or null.",
        messageType: null,
        message: null,
      };
    }

    messageType = MESSAGE_TYPE.RESPONSE;

    // error
  } else if ("error" in message) {
    messageType = MESSAGE_TYPE.ERROR;
  }

  return {
    errorType: null,
    errorMessage: "",
    messageType: messageType,
    message: message,
  };
}

export function encodeRequest(
  method: string,
  id: string | number | null,
  params?: unknown,
): string {
  return JSON.stringify({
    jsonrpc: "2.0",
    id: id,
    method: method,
    params: params,
  });
}

export function encodeNotification(topic: string, data?: unknown): string {
  return JSON.stringify({
    jsonrpc: "2.0",
    method: topic,
    params: data,
  });
}
